src/pip/_internal/index.py

```diff
@@ -270,16 +271,44 @@ class FoundCandidates(object):
     def __init__(
         self,
         candidates,     # type: List[InstallationCandidate]
-        specifier,      # type: specifiers.BaseSpecifier
-        prereleases,    # type: Optional[bool]
+        versions,       # type: Set[_BaseVersion]
+        hashes,         # type: Optional[Hashes]
         sort_key,       # type: Callable[[InstallationCandidate], Any]
     ):
         # type: (...) -> None
         self._candidates = candidates
-        self._specifier = specifier
-        self._prereleases = prereleases
+        self._versions = versions
         self._sort_key = sort_key
 
+    @classmethod
+    def from_specifier(
+        cls,
+        candidates,     # type: List[InstallationCandidate]
+        specifier,      # type: specifiers.BaseSpecifier
+        prereleases,    # type: Optional[bool]
+        hashes,         # type: Optional[Hashes]
+        sort_key,       # type: Callable[[InstallationCandidate], Any]
+    ):
+        # type: (...) -> FoundCandidates
+        versions = set(specifier.filter(
+            # We turn the version object into a str here because otherwise
+            # when we're debundled but setuptools isn't, Python will see
+            # packaging.version.Version and
+            # pkg_resources._vendor.packaging.version.Version as different
+            # types. This way we'll use a str as a common data interchange
+            # format. If we stop using the pkg_resources provided specifier
+            # and start using our own, we can drop the cast to str().
+            [str(c.version) for c in candidates],
+            prereleases=prereleases,
+        ))
+        return cls(candidates, versions, hashes, sort_key)
+
+    def iter_applicable_versions(self):
```

Comment:
I liked the earlier version of this PR more that didn't have this method. The class seems simpler to me if its methods only return `InstallationCandidate` objects, and the `_versions` set is an internal implementation detail that callers don't have to know about.